feat!: replace `chalk` with `styleText` and add `color` to `ResultsMeta`

lumirlumir opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

In this PR, I've migrated from the external `chalk` library to Node.js's native `styleText`.

This PR closes the issue mentioned in #20012.

#### Is there anything you'd like reviewers to focus on?

Closes: #20012

<!-- markdownlint-disable-file MD004 -->

aladdin-add commented on lib/cli-engine/formatters/stylish.js:
```diff
@@ -4,10 +4,11 @@
+const util = require("node:util"),
 	table = require("../../shared/text-table");
 
+/* eslint-disable n/no-unsupported-features/node-builtins -- TODO */
```
should be removed when node v18 was dropped.

mdjermanovic commented:
Will there be any changes in the output of the `stylish` formatter? If not, then since this by itself wouldn't be a breaking change perhaps we could tag it as just `refactor`.

fasttime commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -80,12 +38,14 @@ describe("formatter:stylish", () => {
+			 * assert.strictEqual(chalkStub.reset.callCount, 0);
+			 * assert.strictEqual(chalkStub.yellow.bold.callCount, 0);
+			 * assert.strictEqual(chalkStub.red.bold.callCount, 0);
+			 */
```
To preserve the `callCount` assertion:

```suggestion
			 assert.strictEqual(util.styleText.callCount, 0);
```

or more idiomatically:

```suggestion
			 assert(util.styleText.notCalled);
```

fasttime commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -127,15 +89,17 @@ describe("formatter:stylish", () => {
+				 * assert.strictEqual(chalkStub.reset.callCount, 1);
+				 * assert.strictEqual(chalkStub.yellow.bold.callCount, 0);
+				 * assert.strictEqual(chalkStub.red.bold.callCount, 2);
+				 */
```
To keep the assertions:

```suggestion
			assert(util.styleText.calledWith("reset"));
			assert(util.styleText.neverCalledWith("yellow"));
			assert(util.styleText.calledWith("red"));
```

No trivial way to assert that `["red", "bold"]` was used twice. If the invocation order is assumed, a particular call can be checked by its index:

```js
assert(util.styleText.getCall(4).calledWith(["red", "bold"]));
assert(util.styleText.getCall(5).calledWith(["red", "bold"]));
```

Similarly for the other test cases.

lumirlumir commented on .github/workflows/ci.yml:
```diff
@@ -68,6 +68,7 @@ jobs:
     test_on_node:
         name: Test
         strategy:
+            fail-fast: false
```
Done! https://github.com/eslint/eslint/pull/20227/commits/0e6300c809ad8b073ca07c575c4d9267b011ef1c

lumirlumir commented on tests/tools/check-rule-examples.js:
```diff
@@ -101,7 +101,7 @@ describe("check-rule-examples", () => {
 				"\x1B[0m\x1B[0m\n" +
 				"\x1B[0m\x1B[31m\x1B[1m✖ 18 problems (18 errors, 0 warnings)\x1B[22m\x1B[39m\x1B[0m\n" +
-				"\x1B[0m\x1B[31m\x1B[1m\x1B[22m\x1B[39m\x1B[0m\n";
+				"\x1B[0m\x1B[0m\n";
```
@mdjermanovic I think this part could cause a breaking change.

The counterpart of this line is here:

https://github.com/eslint/eslint/blob/e2516713bc9ae62117da3f490d9cb6a9676f44fe/lib/cli-engine/formatters/stylish.js#L90-L103

---

When using `chalk`, even an empty last line was wrapped with `red` and `bold` ANSI escapes (represented by number pair `[31, 39]` and `[1, 22]`).

Those ANSI escape sequences didn’t affect CLI users because there were no characters to display, so nothing appeared.

However, if someone relied on this output programmatically, this change could be breaking.

---

I could apply `red` and `bold` to the last line again to match the previous behavior, but I feel adding ANSI escapes to an empty line isn’t meaningful.

Which approach is preferred? Would it be fine to remove the `red` and `bold` escapes for this line, or should the previous behavior be preserved?

lumirlumir commented on lib/cli-engine/formatters/stylish.js:
```diff
@@ -98,25 +100,33 @@ module.exports = function (results) {
+				.split("\n")
+				.map(line => util.styleText("reset", line))
+				.join("\n")
+		: "";
```
This is a somewhat interesting internal behavior of `chalk`.

If I change the code to `total > 0 ? styleText("reset", output) : "";`, the tests in `tests/tools/check-rule-examples` fail.

<img width="1540" height="765" alt="Image" src="https://github.com/user-attachments/assets/97e633f4-42dc-416d-a769-8f5f9561d8de" />

---

This is because `chalk.reset()` appears to apply the reset ANSI escape (represented by the `\x1B[0m` sequence) on a per-line basis, while `styleText('reset', ...)` wraps the entire text with a single `reset` ANSI escape.

To mimic `chalk.reset()`'s behavior, this change was unavoidable. I believe `chalk.reset()` wraps the `reset` ANSI escape per line to prevent ANSI escapes from leaking into the next line.

fasttime replied:
Thanks for the detailed explanation. If the output is visually identical for our use case, I think it makes sense to simplify the logic and update `tests/tools/check-rule-examples.js` accordingly. I would still suggest keeping the ANSI escape sequences in that test rather than using `stripVTControlCharacters`, so we will notice in case another update changes the formatting. What do you think?

fasttime replied:
Sorry for being unclear. I was suggesting to call the `styleText` function on the whole `output` for simplicity, and update the tests in `tests/tools/check-rule-examples.js` if necessary. Since the visual effect for terminal users would be the same, this wouldn't account as a breaking change.

fasttime replied:
Chalk applies all styles line by line since v1.1.2. Looking at [05f87e25e108726ee469aa56cc88fcc160a911ed](https://github.com/chalk/chalk/commit/05f87e25e108726ee469aa56cc88fcc160a911ed) and chalk/chalk#92 it seems that this was done to workaround an issue with iTerm2 on Mac. I tried to repro the issue with the current iTerm2 version and chalk v1.1.1 but I didn't manage.

<img width="1364" height="1148" alt="image" src="https://github.com/user-attachments/assets/ed3181f3-6fc8-4ba6-9dab-e78513ba4ed4" />

```js
const chalk = require('chalk');
const text = chalk.bgGreen('hello\n') + 'Lorem ipum dolor sit amet, consectetur adipiscing elit. Vivamus lacinia odio vitae vestibulum vestibulum. Cras venenatis euismod malesuada.';
console.log(text);
```

mdjermanovic replied:
Okay, then one reset for the whole output seems fine.

lumirlumir commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -80,12 +38,10 @@ describe("formatter:stylish", () => {
 
 		it("should not return message", () => {
-			const result = formatter(code);
+			const result = util.stripVTControlCharacters(formatter(code));
```
`chalk` supported disabling ANSI escapes by setting `chalk.level = 0` (previously on line 61). Since `styleText` doesn't provide a way to disable them, I used `util.stripVTControlCharacters` to strip the ANSI escape codes instead.

lumirlumir commented:
This PR is now ready for review!

---

The CI below may fail because it's been running for over 15 minutes; I think it's a transient issue, so re-running it will likely resolve the problem.

<img width="809" height="40" alt="image" src="https://github.com/user-attachments/assets/469aa410-a0d6-43f3-9114-4cea36e66ec2" />

mdjermanovic reviewed (commented):
ESLint CLI officially supports `--no-color` flag:

https://eslint.org/docs/latest/use/command-line-interface#--color-and---no-color

I don't see we're using it anywhere in the eslint code, but when it's passed, the stylish formatter (the version with `chalk`) indeed produces a non-colorized output, so I'm guessing `chalk` itself handles this flag. However, after this change, the output seems to be colorized when `--no-color` is passed.

fasttime commented:
> Looks like `--color`/`--no-color` was handled by https://www.npmjs.com/package/supports-color, a dependency of `chalk`.
> 
> It also supports some environment variables, like `FORCE_COLOR=0`, that will no longer be supported when we switch to native `util.styleText`. So, although this behavior wasn't documented in ESLint, perhaps it's worth marking this change as breaking and add a note about environment variables in the migration guide?
> 
> Regardless, I think we should still support `--color`/`--no-color`.

Good catch on `--color` and `--no-color`. I agree those options should remain supported in v10, and it would be good to add unit tests.

As for marking this as a breaking change, I don't feel strongly if the only difference is the formatting with certain undocumented environment settings.

Maybe it's worth noting that Node.js provides [`process.stdout.hasColors()`](https://nodejs.org/api/tty.html#writestreamhascolorscount-env), which recognizes `FORCE_COLOR=0` along with other environment variables. That API was partly inspired by `supports-color` ([reference](https://github.com/nodejs/node/blob/037a673b9b042ebdc454180a65ec46535c8be86f/lib/internal/tty.js#L120-L122)), though I'm sure there are differences. If replacing `supports-color` is a concern, that's probably the closest built-in equivalent available in the Node.js APIs.

fasttime commented:
> Then, `util.styleText()` probably already uses `process.stdout.hasColors()`

It seems so. When I run the current branch with with `FORCE_COLOR=0`, I don'see any colors in the output. This is in a PowerShell (not sure if there are any differences across different terminals):

<img width="1711" height="345" alt="image" src="https://github.com/user-attachments/assets/ef9fd47e-5b6b-419c-a293-8b00e9b2a51a" />

mdjermanovic commented:
The same difference is observable when the output is redirected to a file:

```bash
npx eslint tests/fixtures/cli/syntax-error.js > output.txt
```

Content of `output.txt` on the `main` branch:

```

C:\projects\eslint\tests\fixtures\cli\syntax-error.js
  1:3  error  Parsing error: Unexpected token }

✖ 1 problem (1 error, 0 warnings)


```

Content of `output.txt` on this PR's branch:

```

C:\projects\eslint\tests\fixtures\cli\syntax-error.js
  1:3  error  Parsing error: Unexpected token }

[31m[1m✖ 1 problem (1 error, 0 warnings)[22m[39m


```

It could be that `util.styleText()` is not stable enough for production use on versions of Node.js supported by ESLint v10.

lumirlumir commented:
Another possible way to fix this problem would be to call `styleText` twice instead of using an array like `[summaryColor, 'bold']`, as shown below:

```js
	if (total > 0) {
		output += `${util.styleText(
			summaryColor,
			util.styleText(
				"bold",
				[
					"\u2716 ",
					total,
					pluralize(" problem", total),
					" (",
					errorCount,
					pluralize(" error", errorCount),
					", ",
					warningCount,
					pluralize(" warning", warningCount),
					")",
				].join(""),
			),
		)}\n`;

		if (fixableErrorCount > 0 || fixableWarningCount > 0) {
			output += `${util.styleText(
				summaryColor,
				util.styleText(
					"bold",
					[
						"  ",
						fixableErrorCount,
						pluralize(" error", fixableErrorCount),
						" and ",
						fixableWarningCount,
						pluralize(" warning", fixableWarningCount),
						" potentially fixable with the `--fix` option.",
					].join(""),
				),
			)}\n`;
		}
	}
```	

---

When I run the following:

```sh
npx eslint tests/fixtures/cli/syntax-error.js > output.txt
```

I can get the intended result as follows:

```txt

D:\git\eslint\tests\fixtures\cli\syntax-error.js
  1:3  error  Parsing error: Unexpected token }

✖ 1 problem (1 error, 0 warnings)

```

---

How about this approach? If that works, I can update the logic and add more test cases to cover it.

lumirlumir commented:
@mdjermanovic @fasttime 

I've updated the logic and added tests for the `FORCE_COLOR` environment variable. ( This is an interim result, so adding support for `--color` and `--no-color` is my next step :) )

While working on this, I have two questions.

## 1. First

According to the Node.js documentation, there are two additional environment variables (`NO_COLOR` and `NODE_DISABLE_COLORS`) that can disable ANSI styling besides `FORCE_COLOR`.

https://nodejs.org/api/tty.html#writestreamgetcolordepthenv

<img width="806" height="61" alt="image" src="https://github.com/user-attachments/assets/77b6e6c1-8292-4c90-bde5-059c9061e79e" />

So, the following test code does not include ANSI escape sequences.

```js
import { styleText } from "node:util";

process.env.FORCE_COLOR = "0";

console.log(styleText("red", "This is a styled text"));
```

```js
import { styleText } from "node:util";

process.env.NO_COLOR = true;

console.log(styleText("red", "This is a styled text"));
```

```js
import { styleText } from "node:util";

process.env.NODE_DISABLE_COLORS = true;

console.log(styleText("red", "This is a styled text"));
```

---

However, when I run the following code with `NODE_DISABLE_COLORS=true` and `NO_COLOR=true`, `chalk` ignores those environment variables and still shows colored output:

```js
/* test.mjs */

// Run it using `NO_COLOR=true node test.mjs` or `NODE_DISABLE_COLORS=true node test.mjs`

import chalk from "chalk";

console.log(chalk.red("This is a styled text"));
```

How should we handle the `NO_COLOR` and `NODE_DISABLE_COLORS` environment variables? Would that be considered a new feature in this change?

Also, would it make sense to add test cases for these two environment variables as well?

## 2. Second

Should we update the documentation to describe the `FORCE_COLOR`, `NO_COLOR`, and `NODE_DISABLE_COLORS` environment variables? If so, where would be the best place to include those descriptions?

---

References:

- `NODE_DISABLE_COLORS`: https://nodejs.org/api/cli.html#node_disable_colors1
- `NO_COLOR`: https://nodejs.org/api/cli.html#no_colorany

lumirlumir commented:
I think I've resolved all comments in the threads.

> As for --color/--no-color, I think we'll need to check CLI arguments in the formatter (perhaps in the future consider passing this info to formatters), and then:

Regarding `--color`/`--no-color`, per the earlier suggestion, the flag information is now passed to formatters via the `ResultsMeta` type.

I've renamed the PR prefix to `feat` to reflect this change and added tests to ensure this behavior.

mdjermanovic commented on lib/cli.js:
```diff
@@ -454,14 +454,24 @@ const cli = {
+		if (options.color !== void 0) {
+			debug(`Color setting for output: ${options.color}`);
+			resultsMeta.color = options.color;
+		}
```
@nzakas what do you think about this?

lumirlumir replied:
Are there any other places I should update? From the discussion, I'm not sure whether I need to update any additional code. Any clarification would be appreciated.

lumirlumir commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -10,63 +10,169 @@
+		});
+
+		it("`FORCE_COLOR=0` should disable colors", () => {
+			process.env.FORCE_COLOR = 0;
```
`process.env.FORCE_COLOR=0` and `process.env.FORCE_COLOR="0"` work exactly the same, so there's no need to update it.

lumirlumir commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -10,63 +10,169 @@
+		});
+
+		it("`FORCE_COLOR=2` should enable colors", () => {
+			process.env.FORCE_COLOR = 2;
```
Same with https://github.com/eslint/eslint/pull/20227#discussion_r2602259578

lumirlumir commented on tests/lib/cli-engine/formatters/stylish.js:
```diff
@@ -10,63 +10,169 @@
+		});
+
+		it("`color: true` should ignore environment variable", () => {
+			process.env.FORCE_COLOR = 0;
```
Same with https://github.com/eslint/eslint/pull/20227#discussion_r2602259578

fasttime reviewed (commented):
Also here:

https://github.com/eslint/eslint/blob/ca8a4faff717d93778d44343860bbb04fc8e6b9d/lib/types/index.d.ts#L1317-L1319

lumirlumir commented on tests/lib/cli.js:
```diff
@@ -253,6 +253,47 @@ describe("cli", () => {
+					const { metadata } = JSON.parse(log.info.args[0][0]);
+
+					assert.notProperty(metadata, "color");
+				});
```
WIP

I created `context.js` as a fixture formatter that simply returns the `context`.

I also updated the `maxWarningsExceeded` test to reflect this change.

The change is here:

lumirlumir reviewed (commented):
I think I've resolved all of the comments.

Another pair of eyes on it would be helpful!

fasttime commented on docs/src/use/migrate-to-10.0.0.md:
```diff
@@ -168,6 +169,26 @@ Here, `[[:upper:]]` is a POSIX character class that matches uppercase letters in
+
+    Please note that the [`FORCE_COLOR`](https://nodejs.org/api/cli.html#force_color1-2-3) environment variable is still supported to force-enable colors.
+
+2. Second, `--color` and `--no-color` CLI flags now have higher precedence than environment variables when determining whether to use colorized output. This change ensures that explicit user preferences via CLI flags are prioritized. However, if neither flag is provided, environment variables will be considered as before.
```
What do you think about swapping points 1 and 2 to reflect the fact that `--color` and `--no-color` are evaluated first, and Node.js defaults are applied only when neither flag is specified?

fasttime reviewed (approved):
LGTM, thanks!
